liquibase/liquibase
# Avoid NullPointerException when customChange has no "class" attribute

## Environment
**Liquibase Version**: all

**Liquibase Integration & Version**: all

**Liquibase Extension(s) & Version**: all

**Database Vendor & Version**: all

**Operating System Type & Version**: all

## Pull Request Type


* [ ] Enhancement/New feature (non-breaking change which adds functionality)

## Description
If a customChange node is missing the "class" attribute (either by typo, or mis-indentation) a NullPointerException is thrown:

```
Caused by: java.lang.NullPointerException
    at java.lang.Class.forName0(Native Method) ~[?:?]
    at java.lang.Class.forName(Class.java:398) ~[?:?]
    at liquibase.change.custom.CustomChangeWrapper.customLoadLogic(CustomChangeWrapper.java:335) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]
    at liquibase.change.AbstractChange.load(AbstractChange.java:719) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]
    at liquibase.change.custom.CustomChangeWrapper.load(CustomChangeWrapper.java:312) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]
    at liquibase.changelog.ChangeSet.toChange(ChangeSet.java:480) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]
    at liquibase.changelog.ChangeSet.handleChildNode(ChangeSet.java:413) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]   
    at liquibase.changelog.ChangeSet.handleChildNode(ChangeSet.java:409) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]   
    at liquibase.changelog.ChangeSet.load(ChangeSet.java:337) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]
    at liquibase.changelog.DatabaseChangeLog.createChangeSet(DatabaseChangeLog.java:606) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]
    at liquibase.changelog.DatabaseChangeLog.handleChildNode(DatabaseChangeLog.java:336) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]
    at liquibase.changelog.DatabaseChangeLog.load(DatabaseChangeLog.java:305) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?] 
    at liquibase.parser.core.yaml.YamlChangeLogParser.parse(YamlChangeLogParser.java:77) ~[usafe-cli-1.0.1-SNAPSHOT-all.jar:?]
```

## Steps To Reproduce
Create a customChange node with no class attribute, and you will see this exception.

## Actual Behavior
Check description.

## Expected/Desired Behavior
A nicer exception should appear with a better error message.

## Fast Track PR Acceptance Checklist:




* [x] Build is successful and all new and existing tests pass
* [ ] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
* [ ] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
* [ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: Stand alone PR
* Test Results: https://github.com/liquibase/liquibase/pull/1648/checks

#### Testing
* Steps to Reproduce: See above (use yaml format)
* Guidance:
  * Impact: Just adds a check for a missing argument

#### Dev Verification
Code review and ensured new automated test is good

### Manual Tests

_Verify update shows correct error message when class attribute is not specified._

- create a changelog with customChange changeset, but do not specify class attribute

- `liquibase --log-level INFO update --changelog-file changelog.yaml`

- `Custom change node has no 'class' attribute!` error message is shown

_Verify update is successful when class attribute is specified._

- specify 'class' attribute in your changelog

- `liquibase update --changelog-file changelog.yaml`

- update is successful


┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-1144) by [Unito](https://www.unito.io)
